balancer/pickfirst/pickfirstleaf/pickfirstleaf_ext_test.go

```diff
@@ -851,6 +850,253 @@ func (s) TestPickFirstLeaf_EmptyAddressList(t *testing.T) {
+	testutils.AwaitState(ctx, t, cc, connectivity.TransientFailure)
+}
+
+// TestPickFirstLeaf_HappyEyeballs_TriggerConnectionDelay verifies that
+// pickfirst attempts to connect to the second backend once the happy eyeballs
+// timer expires.
+func (s) TestPickFirstLeaf_HappyEyeballs_TriggerConnectionDelay(t *testing.T) {
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+
+	timerCh := make(chan struct{})
+	originalTimer := pfinternal.TimeAfterFunc
+	pfinternal.TimeAfterFunc = func(_ time.Duration, f func()) *time.Timer {
+		// Set a really long expiration to prevent it from triggering
+		// automatically.
+		timer := time.AfterFunc(time.Hour, f)
+		go func() {
+			select {
+			case <-ctx.Done():
+			case <-timerCh:
+			}
+			timer.Reset(0)
+		}()
+		return timer
+	}
+
+	defer func() {
+		pfinternal.TimeAfterFunc = originalTimer
+	}()
+
+	defer func() {
+		pfinternal.TimeAfterFunc = originalTimer
+	}()
+
+	dialer := testutils.NewBlockingDialer()
+	cc, rb, bm := setupPickFirstLeaf(t, 2,
+		grpc.WithDefaultServiceConfig(fmt.Sprintf(`{"loadBalancingConfig": [{"%s":{}}]}`, pickfirstleaf.Name)),
+		grpc.WithContextDialer(dialer.DialContext),
+	)
+	addrs := bm.resolverAddrs()
+	holds := bm.holds(dialer)
+	rb.UpdateState(resolver.State{Addresses: addrs})
+	cc.Connect()
+
+	testutils.AwaitState(ctx, t, cc, connectivity.Connecting)
+
+	// Verify that the first server is contacted.
+	if got, want := holds[0].Wait(ctx), true; got != want {
+		t.Fatalf("hold[%q].Wait() = %t, want %t", addrs[0], got, want)
+	}
+
+	if got, want := holds[1].IsStarted(), false; got != want {
+		t.Fatalf("holds[%q].IsStarted() = %t, want %t", addrs[1], got, want)
+	}
+
+	timerCh <- struct{}{}
+
+	// Second connection attempt is successful.
+	if got, want := holds[1].Wait(ctx), true; got != want {
+		t.Fatalf("hold[%q].Wait() = %t, want %t", addrs[1], got, want)
+	}
+	holds[1].Resume()
+	testutils.AwaitState(ctx, t, cc, connectivity.Ready)
+}
+
+// TestPickFirstLeaf_HappyEyeballs_TFThenTimerFires tests the pickfirst balancer
+// by causing a SubConn to fail and then jumping to the 3rd SubConn after the
+// happy eyeballs timer expires.
+func (s) TestPickFirstLeaf_HappyEyeballs_TFThenTimerFires(t *testing.T) {
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+
+	timerMu := sync.Mutex{}
+	timerCh := make(chan struct{})
+	originalTimer := pfinternal.TimeAfterFunc
+	pfinternal.TimeAfterFunc = func(_ time.Duration, f func()) *time.Timer {
+		// Set a really long expiration to prevent it from triggering
+		// automatically.
+		timer := time.AfterFunc(time.Hour, f)
+		timerMu.Lock()
+		ch := timerCh
+		timerMu.Unlock()
+		go func() {
+			select {
+			case <-ctx.Done():
+			case <-ch:
+			}
+			timer.Reset(0)
+		}()
+		return timer
+	}
+
+	defer func() {
+		pfinternal.TimeAfterFunc = originalTimer
+	}()
+
+	dialer := testutils.NewBlockingDialer()
+	cc, rb, bm := setupPickFirstLeaf(t, 3,
```

Comment:
Many of the comments on the first test, applies to the second and third one as well. Please do take care of all three tests. Thanks.